src/integer/mat_z/arithmetic/div.rs

```diff
@@ -0,0 +1,292 @@
+    ///
+    /// let mat2 = &mat1 / &integer;
+    /// ```
+    ///
+    /// # Panics ...
+    /// - if the divisor is `0`.
+    fn div(self, divisor: &Z) -> Self::Output {
+        if divisor == &Z::ZERO {
+            panic!("DivisionByZero: tried to divide {} by zero", self);
+        }
+
+        let mut out = MatQ::new(self.get_num_rows(), self.get_num_columns());
+        unsafe {
+            fmpq_mat_set_fmpz_mat_div_fmpz(&mut out.matrix, &self.matrix, &divisor.value);
+        }
+        out
+    }
+}
+
+arithmetic_trait_borrowed_to_owned!(Div, div, MatZ, Z, MatQ);
+arithmetic_trait_mixed_borrowed_owned!(Div, div, MatZ, Z, MatQ);
+
+impl MatZ {
+    /// Implements division for a [`MatZ`] matrix by a [`Z`] integer.
+    ///
+    /// Parameters:
+    /// - `divisor`: specifies the divisor by which the matrix is divided
+    ///
+    /// Returns the product of `self` and `other` as a [`MatZ`] or panics if
+    /// the divisor is 0.
+    ///
+    /// # Safety
+    /// The divisor MUST exactly divide each element in the matrix.
+    /// If this is not the case, the result can contain arbitrary values
+    /// which can depend on the location in memory.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer::{MatZ, Z};
+    /// use std::str::FromStr;
+    ///
+    /// let mut mat1 = MatZ::from_str("[[3,6],[9,27]]").unwrap();
+    /// let integer = Z::from(3);
+    ///
+    /// let mat2 = unsafe{ mat1.div_exact(&integer)};
+    /// ```
+    pub unsafe fn div_exact(mut self, divisor: impl Into<Z>) -> MatZ {
+        let divisor: Z = divisor.into();
+
+        fmpz_mat_scalar_divexact_fmpz(&mut self.matrix, &self.matrix, &divisor.value);
+
+        self
+    }
+
+    /// Implements division for a [`MatZ`] matrix by a [`Z`] integer.
+    ///
+    /// Parameters:
+    /// - `divisor`: specifies the divisor by which the matrix is divided
+    ///
+    /// Returns the product of `self` and `other` as a [`MatZ`] or panics if
+    /// the divisor is 0.
+    ///
+    /// # Safety
+    /// The divisor MUST exactly divide each element in the matrix.
+    /// If this is not the case, the result can contain arbitrary values
+    /// which can depend on the location in memory.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer::{MatZ, Z};
+    /// use std::str::FromStr;
+    ///
+    /// let mat1 = MatZ::from_str("[[3,6],[9,27]]").unwrap();
+    /// let integer = Z::from(3);
+    ///
+    /// let mat2 = unsafe{&mat1.div_exact_ref(&integer)};
+    /// ```
+    #[allow(clippy::missing_safety_doc)]
+    pub unsafe fn div_exact_ref(&self, divisor: impl Into<Z>) -> MatZ {
+        let divisor: Z = divisor.into();
+
+        let mut out = MatZ::new(self.get_num_rows(), self.get_num_columns());
+        fmpz_mat_scalar_divexact_fmpz(&mut out.matrix, &self.matrix, &divisor.value);
+
+        out
+    }
+}
+
+#[cfg(test)]
+mod test_div_exact {
+    use super::*;
+    use std::str::FromStr;
+
```

Comment:
1. Add a test to showcase the different types with which the function can be called with. e.g. Z,  Modulus, rust integers (owned and borrowed).
Then, you can also replace borrowed_ , owned_ , and mixed_correctness with one test e.g. small_numbers.

2. Also, you can use the rust integers directly in the tests below to make them ore simple instead of creating a Z first. If you stick with the variable for the divisor, you should also call it "divisor" and not just "integer".

3. Add (one or more) tests where negative matrix entries are divided.

4. Test with negative divisor.

Reply:
Some clarifications of what i meant with this:
> Add a test to showcase the different types with which the function can be called with. e.g. Z, Modulus, rust integers (owned and borrowed).
Then, you can also replace borrowed_ , owned_ , and mixed_correctness with one test e.g. small_numbers.

I expected you to create one test to test all the different type combinations. We already have 30 tests like that called `availability`. This way the code should become much shorter because you don't have to initialize mat1, mat2, mat_cmp all the time.

Additionally, you don't have to create a `mat2` if you don't shadow `mat1` and just clone when you call the function or by first using `div_exact_ref` and then `div_exact`.
e.g. for availability:
``` rust
let mat = MatZ::from_str(...).unwrap();

// div_exact_ref
let _ = unsafe{ mat.div_exact_ref(3i64)};
let _ = unsafe{ mat.div_exact_ref(3u32)};
let _ = unsafe{ mat.div_exact_ref(Z::from(3))};
[... Also for the borrowed versions and maybe some more different rust integer types]

/// div_exact
let _ = unsafe{ mat.clone().div_exact(3i64) };
let _ = unsafe{ mat.clone().div_exact(3iu32) };
[... Also for the borrowed versions and maybe some more different rust integer types]
```

For the other tests:
```rust
let mat = [...]

let mat_divided_1 = unsafe{ mat.div_exact_ref(3)};
let mat_divided_2 = unsafe{ mat.div_exact(3) };

let mat_cmp = [...]
assert_eq!(mat_divided_1, mat_cmp);
assert_eq!(mat_divided_2, mat_cmp);
```

> 2. Also, you can use the rust integers directly in the tests below to make them ore simple instead of creating a Z first. If you stick with the variable for the divisor, you should also call it "divisor" and not just "integer".

Not sure why you did not do this.

> 3. Add (one or more) tests where negative matrix entries are divided.
> 4. Test with negative divisor.

Do the negative test in a dedicated test function. `types_correctness` and `different_dimensions_correctness` do not sound like checking negative divisors/values and the comment also do not mention it.
Furthermore, also test with large negative divisors and large negative values.
You can combine 3. and 4. within the same test.

Additionally, you could consider removing the `different_dimensions_correctness` test and just use different dimensions throughout the other tests.

Reply:
I would call it integer because we call the functionality "division by integer"